apis/user/v1/invitation_types.go

```diff
@@ -0,0 +1,137 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/apiserver-runtime/pkg/builder/resource"
+
+	"github.com/appuio/control-api/apiserver/secretstorage/status"
+)
+
+const (
+	// ConditionRedeemed is set when the invitation has been redeemed
+	ConditionRedeemed = "Redeemed"
+	// ConditionEmailSent is set when the invitation email has been sent
+	ConditionEmailSent = "EmailSent"
+)
+
+// +kubebuilder:object:root=true
+
+// Invitation is a representation of an APPUiO Cloud Invitation
+type Invitation struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec holds the desired invitation state
+	Spec InvitationSpec `json:"spec,omitempty"`
+	// Status holds the invitation specific status
+	Status InvitationStatus `json:"status,omitempty"`
+}
+
+// InvitationSpec defines the desired state of the Invitation
+type InvitationSpec struct {
+	// Note is a free-form text field to add a note to the invitation
+	Note string `json:"note,omitempty"`
+	// Email is the email address of the invited user, used to send the invitation
+	Email string `json:"email,omitempty"`
+	// TargetRefs is a list of references to the target resources
+	TargetRefs []TargetRef `json:"targetRefs,omitempty"`
+}
+
+// TargetRef is a reference to a target resource
+type TargetRef struct {
```

Comment:
There's probably a good reason why we define `TargetRef` ourselves, but I'm wondering if we could just use the standard [`ObjectReference`](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/object-reference/) type instead.

Reply:
I think the reason was looking into the `RoleBinding` subjects instead of other references. Those don't have the right documentation when generated.

Using `ObjectReference` seems like a good idea but will introduce more validation overhead since there are more fields to respect and well most likely never use any of them.

Reply:
I'm fine with having our own ref type especially since that will give us flexibility if we figure out that we need something that we can't get from `ObjectReference`. Additionally, there seems to be no clear consensus even in the core K8s types on whether you're supposed to use `LocalObjectReference`/`ObjectReference` vs your own ref types